runtime/permissions/mod.rs

```diff
@@ -226,22 +228,267 @@ impl AsRef<str> for EnvVarName {
   }
 }
 
+pub trait Descriptor: Eq + Clone {
+  fn flag_name() -> &'static str;
+  fn name(&self) -> Cow<str>;
+  // By default, specifies no-stronger-than relationship.
+  // As this is not strict, it's only true when descriptors are the same.
+  fn stronger_than(&self, other: &Self) -> bool {
+    self == other
+  }
+}
+
 #[derive(Clone, Debug, Eq, PartialEq)]
-pub struct UnaryPermission<T: Eq + Hash> {
-  pub name: &'static str,
-  pub description: &'static str,
-  pub global_state: PermissionState,
+pub struct UnaryPermission<T: Descriptor + Hash> {
+  pub granted_global: bool,
   pub granted_list: HashSet<T>,
-  pub denied_list: HashSet<T>,
+  pub flag_denied_global: bool,
+  pub flag_denied_list: HashSet<T>,
+  pub prompt_denied_global: bool,
+  pub prompt_denied_list: HashSet<T>,
   pub prompt: bool,
 }
 
+impl<T: Descriptor + Hash> Default for UnaryPermission<T> {
+  fn default() -> Self {
+    UnaryPermission {
+      granted_global: Default::default(),
+      granted_list: Default::default(),
+      flag_denied_global: Default::default(),
+      flag_denied_list: Default::default(),
+      prompt_denied_global: Default::default(),
+      prompt_denied_list: Default::default(),
+      prompt: Default::default(),
+    }
+  }
+}
+
+impl<T: Descriptor + Hash> UnaryPermission<T> {
+  fn check_desc(
+    &mut self,
+    desc: &Option<T>,
+    assert_non_partial: bool,
+    api_name: Option<&str>,
+    get_display_name: impl Fn() -> Option<String>,
+  ) -> Result<(), AnyError> {
+    let (result, prompted, is_allow_all) =
+      self.query_desc(desc, Some(assert_non_partial)).check2(
+        T::flag_name(),
+        api_name,
+        || match get_display_name() {
+          Some(display_name) => Some(display_name),
+          None => desc.as_ref().map(|d| format!("\"{}\"", d.name())),
+        },
+        self.prompt,
+      );
+    if prompted {
+      if result.is_ok() {
+        if is_allow_all {
+          self.insert_granted(None);
+        } else {
+          self.insert_granted(desc.clone());
+        }
+      } else {
+        self.insert_prompt_denied(desc.clone());
+      }
+    }
+    result
+  }
+
+  /// `allow_partial` prescribes how to treat a permission which is partially
+  /// denied due to a `--deny-*` flag affecting a subscope of the queried
+  /// permission. `Some(true)` means treat as granted, `Some(false)` means treat
+  /// as denied and `None` means report as `PermissionState::GrantedPartial`.
+  /// `Some(true)` is used in place of `None` when we don't want to wastefully
+  /// check for partial denials when, say, checking read access for a file.
+  fn query_desc(
+    &self,
+    desc: &Option<T>,
+    allow_partial: Option<bool>,
```

Comment:
It might be good to make this an enum with three options so it's more explanatory here and at the call sites.